### ISSUE TITLE
annotation & property of completion item has no effect when helm-completion-style is set

### ISSUE THREAD

**Issue by twlz0ne** [emacs-helm/helm#2378]
## Expected behavior

annotation & property has an effect

![emacs-helm-annotation-has-an-effect](https://user-images.githubusercontent.com/2653486/92192870-852dce00-ee99-11ea-8d59-361c8870d98d.png)

## Actual behavior

annotation & property  has no effect

![emacs-helm-annotiation-has-no-affect](https://user-images.githubusercontent.com/2653486/92192758-49930400-ee99-11ea-86d8-887004cd444e.png)

## Steps to reproduce (recipe)

``` elisp
emacs-nightly -nw -Q -l test-helm-completion-annotation-issue.el
```

<details>
<summary><code>test-helm-completion-annotation-issue.el</code></summary>

``` elisp
;;; Usage: /path/to/emacs -nw -Q -l /path/to/test-helm-completion-annotation-issue.el
;;; Date: 2020-09-04 10.09.29
(toggle-debug-on-error)

(add-to-list 'load-path (expand-file-name "~/repos/emacs-async"))
(add-to-list 'load-path (expand-file-name "~/repos/emacs-helm"))

(require 'helm)
(require 'helm-config)

(defun my-annotation-function (s)
  (let ((item (member s '("bar"))))
    (when item " *")))

(defun test()
  (interactive)
  (let ((completion-extra-properties '(:annotation-function my-annotation-function)))
    (completing-read "Prompt: "
                     (mapcar (lambda (item)
                               (cons (propertize (car item) 'face 'error)
                                     (cdr item)))
                             '(("foo" 1)
                               ("bar" 2)
                               ("qux" 3))))))

(add-hook 'after-init-hook
          (lambda ()
            (setq helm-completion-style 'helm-fuzzy) ;;<-----
            (helm-mode 1)
            (test)))

(run-hooks 'after-init-hook)
;;; test-helm-completion-annotation-issue.el ends here
```

</details>

## Backtraces if any (`M-x toggle-debug-on-error`)

No

## Describe versions of Helm, Emacs, operating system, etc.

- helm r378fe0d2 (2020-09-03)
- Emacs 28.0 (2020-08-28)
- macOS 10.12.6

## Are you using `emacs-helm.sh` to reproduce this bug? (yes/no):

No

## Are you using Spacemacs? (yes/no):

No

**Comment by thierryvolpiatto** [comment]
twlz0ne <notifications@github.com> writes:

> Expected behavior
>
> annotation & property has an effect
>
> emacs-he
>
> Actual behavior
>
> annotation & property has no effect
>
> emacs-he
>
> Steps to reproduce (recipe)
>
> emacs-nightly -nw -Q -l test-helm-completion-annotation-issue.el
>
> (add-hook 'after-init-hook
>           (lambda ()
>             (setq helm-completion-style 'helm-fuzzy) ;;<-----

This can't work with `helm-fuzzy`, use `emacs` as
`helm-completion-style` and if you want fuzzy/flex completion use 'flex
(emacs-27+) in completion-styles:

    (setq completion-styles ' (flex))

Thanks.

-- 
Thierry

**Comment by twlz0ne** [comment]
> This can't work with `helm-fuzzy`, use `emacs` as `helm-completion-style` and if you want fuzzy/flex completion use 'flex (emacs-27+) in completion-styles: (setq completion-styles ' (flex))

Sorry for not reading the document.  I use `helm-fuzzy` for a long time and never noticed anything wrong with it.

The `'(flex)` does solve the annotation issue, but it seems to bring a new problem:

![emacs-27-helm-completion-style-_flex_](https://user-images.githubusercontent.com/2653486/92256075-ceb40280-ef05-11ea-953d-fa02d39bec92.png)

It makes me always kill the wrong buffer.  I found that was introduced in https://github.com/emacs-helm/helm/commit/070adab3a18def6f9a602cad98ba9197e106606f

**Comment by manuel-uberti** [comment]
Could you try from `emacs-helm.sh`? I am not able to reproduce from there:

![Screenshot from 2020-09-04 18-48-06](https://user-images.githubusercontent.com/8199224/92264871-330f9b80-eedf-11ea-8672-85284ca1c145.png)

**Comment by thierryvolpiatto** [comment]
twlz0ne <notifications@github.com> writes:

>     This can't work with helm-fuzzy, use emacs as helm-completion-style and if you want fuzzy/flex completion use 'flex (emacs-27+) in completion-styles: (setq completion-styles ' (flex))
>    
> Sorry for not reading the document. I use helm-fuzzy for a long time and never noticed anything wrong with it.
>
> The '(flex) does solve the annotation issue, but it seems to bring a new problem:
>
> emacs-27
>
> It makes me always kill the wrong buffer. I found that was introduced in 070adab

Thanks to catch this.
I have commited a fix which is not perfect but make at least kill-buffer
and switch-to-buffer behaving correctly.
read-buffer is still wrong as now the current buffer is missing, but for
now it doesn't harm as most higher level functions handle this
(i.e. adding current buffer on top of completion list).

-- 
Thierry

**Comment by twlz0ne** [comment]
> Could you try from `emacs-helm.sh`?

![emacs-helm-completion-style-_flex_-emacs-helm-sh](https://user-images.githubusercontent.com/2653486/92276470-99b6a880-ef23-11ea-8aa6-e013bae49f6c.png)

`emacs-helm.sh` still get the wrong buffer on top.

**Comment by thierryvolpiatto** [comment]
twlz0ne <notifications@github.com> writes:

>     Could you try from emacs-helm.sh?
>    
> emacs-he
>
> emacs-helm.sh still get the wrong buffer on top.

I think you are miscommunicating with Manuel here ;-), IIUC he means
reproduction of the annotation issue (initial bug report) from
emacs-helm.sh, but I think this is now fixed.

I fixed the issue (probably we should start a new bug report as it is
confusing with the initial bug report) for kill-buffer and
switch-to-buffer, but I don't know how to fix globally read-buffer for
now, i.e. I don't know yet how to pass minibuffer-completion-table to it.

-- 
Thierry
